File: luiz/ba1j.py
#! /usr/bin/env python

from collections import defaultdict

from ba1c import revc
from ba1i import neighbors


def frequent_words_with_mismatches_and_rc(text, k, distance):
    frequent_patterns = set()
    kmers = defaultdict(int)
    max_freq = 0

    for i in range(len(text) - k):
        for pattern in neighbors(text[i : i + k], distance):
            kmers[pattern] += 1
            rc = revc(pattern)

            current_freq = kmers[pattern] + kmers[rc]
            if current_freq == max_freq:
                frequent_patterns.add(pattern)
                frequent_patterns.add(rc)
            elif current_freq > max_freq:
                frequent_patterns = set([pattern, rc])
                max_freq = current_freq

    return frequent_patterns


assert frequent_words_with_mismatches_and_rc(
    "ACGTTGCATGTCGCATGATGCATGAGAGCT", 4, 1
) == set("ATGT ACAT".split())


if __name__ == "__main__":
    with open("data/rosalind_ba1j.txt", "r") as dataset:
        text = dataset.readline().rstrip()
        k, d = [int(i) for i in dataset.readline().rstrip().split()]
    print(*frequent_words_with_mismatches_and_rc(text, k, d))

File: luiz/src/bin/ba1j.rs
use std::collections::{HashMap, HashSet};
use std::env;
use std::error::Error;
use std::fs;
use std::str;

use bioinformatics_algorithms::{neighbors, revc};

pub fn frequent_words_with_mismatches_and_rc(text: &[u8], k: usize, d: usize) -> HashSet<String> {
    let mut kmers = HashMap::new();
    let mut frequent = HashSet::new();
    let mut max = 0;

    text.windows(k)
        .map(|x| {
            for pattern in neighbors(str::from_utf8(x).unwrap(), d) {
                let rc_str = revc(&pattern);
                let mut current_freq = 0;
                {
                    let kmer = kmers.entry(pattern.clone()).or_insert(0);
                    *kmer += 1;

                    current_freq += *kmer;
                }

                if let Some(&rc_c) = kmers.get(&rc_str) {
                    current_freq += rc_c
                };

                if current_freq == max {
                    frequent.insert(pattern);
                    frequent.insert(rc_str);
                } else if current_freq > max {
                    max = current_freq;
                    frequent.clear();
                    frequent.insert(pattern);
                    frequent.insert(rc_str);
                }
            }
        })
        .count();

    frequent
}

fn main() -> Result<(), Box<dyn Error>> {
    let input: String = env::args()
        .nth(1)
        .unwrap_or("data/rosalind_ba1j.txt".into());
    let data = fs::read_to_string(input)?;
    let mut lines = data.lines();

    let text = lines.next().unwrap();
    let mut ints = lines
        .next()
        .unwrap()
        .split(' ')
        .map(|x| x.parse().unwrap())
        .take(2);
    let k = ints.next().unwrap();
    let d = ints.next().unwrap();

    for word in frequent_words_with_mismatches_and_rc(text.as_bytes(), k, d) {
        print!("{} ", word);
    }
    println!();

    Ok(())
}
